apps/pipelines/nodes/nodes.py

```diff
@@ -148,40 +213,54 @@ def _create_model_from_data(value_data, model_name: str):
         return schema
 
 
-class UpdateParticipantMemory(PipelineNode):
-    """A simple component to merge the the input data into the participant's memory data. If key_name is specified
-    the input data will be merged with `key_name` as the key name.
-    """
+class ExtractStructuredData(ExtractStructuredDataNodeMixin, LLMResponse):
+    __human_name__ = "Extract Structured Data"
+    data_schema: str
+
 
-    __human_name__ = "Update participant memory"
+class ExtractParticipantData(ExtractStructuredDataNodeMixin, LLMResponse):
+    __human_name__ = "Extract Participant Data"
+    data_schema: str
     key_name: str | None = None
 
-    def _process(self, state: PipelineState) -> PipelineState:
-        """Strategy:
-        - If the input is a string or list, a `key_name` must be present to be valid JSON.
-        - If `key_name` is present, we create a new dictionary with `input` as the value to `key_name`.
-        - Merge this new dictionary with the current participant data, essentially overriding existing data
-        """
-        new_data = state["messages"][-1]
+    def get_json_schema(self):
+        return ExtractParticipantData.to_json_schema(json.loads(self.data_schema))
+
+    def get_reference_data(self, state) -> dict:
         session = state["experiment_session"]
+        data = session.get_participant_data()
+        if self.key_name:
+            return data.get(self.key_name, {})
+        return data
 
+    def update_reference_data(self, new_data: dict, reference_data: dict) -> dict:
         if isinstance(new_data, str) or isinstance(new_data, list):
             if not self.key_name:
                 raise KeyError("A key is expected for a string or list value.")
 
+        if isinstance(new_data, str):
+            return new_data
+
+        if isinstance(new_data, list):
+            return new_data
```

Comment:
it will be interesting to see if it only returns NEW items or if it includes existing items as well

Reply:
I just realized this will never happen actually. Long story short, `new_data` can never be a pure list. It must always be a dict. A key inside the dict can be a list, but only a list of objects. That's currently a limitation. This means that if the existing data looks like
```python
{"some_key": ["val1", "val2", "val3"]}
```

but we want to extract
```python
{"some_key": [{"key1": "val"}]}
```

this will override the original list. I guess if the LLM cannot figure out what the items in the original list means, it will probably discard it. I'll have to think a bit more around this limitation.

Reply:
easiest solution is to not allow updating keys with mismatched types